src/cloudflare/internal/vectorize-api.ts

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+/// <reference path="./vectorize.d.ts" />
+
+interface Fetcher {
+  fetch: typeof fetch;
+}
+
+enum Operation {
+  INDEX_GET,
+  VECTOR_QUERY,
+  VECTOR_INSERT,
+  VECTOR_UPSERT,
+  VECTOR_GET,
+  VECTOR_DELETE,
+}
+
+class VectorSearch {
+  public static async send(
+    fetcher: Fetcher,
+    operation: Operation,
+    endpoint: string,
+    init: RequestInit
+  ): Promise<Response> {
+    const res = await fetcher.fetch(
+      `http://vector-search/binding/${endpoint}`,
+      init
+    );
+    if (res.status !== 200) {
+      try {
+        const err = (await res.json()) as VectorizeError;
+        throw new Error(
+          `${Operation[operation]}_ERROR ${
+            typeof err?.code === "number" ? ` (code = ${err.code})` : ""
+          }: ${err.error}`,
+          {
+            cause: new Error(err.error),
```

Comment:
One quick warning here in that currently our internal error handling mechanisms (e.g. the part that maps JavaScript errors into KJ Exceptions and back) currently is not capable of preserving the `cause`. Depending on how this error is expected to be used, it is possible that detail can be lost in some cases.